### PR TITLE
Initial work on fair share policy

### PR DESCRIPTION
**Pull request by andreasDalsgaard**
Spread load over multiple sources to avoid to filling up one source before using another source. E.g., if an application create several files but wait before starting to write, with mfs all files will be placed on one source. The goal of the fair policy is to spread files created evenly among sources which still have free space. Evenly, in the sense that if one have two sources:
A: having 20 gb free space
and
B: having 10 gb free space.
Then if one create 30 files, then 20 files will be created on source A and 10 files on source B. Since this is faster at spreading files over multiple disks, the bandwidth of both disks will be better utilised, compared to mfs.

### CODE REVIEW

**Comment by trapexit** [comment]
Could you please describe the algorithm in plain terms? "fair" is not particularly descriptive.

**Comment by andreasDalsgaard** [comment]
> Could you please describe the algorithm in plain terms? "fair" is not particularly descriptive. What does "overfilling" mean? And in what way it "should give better performance compared to mfs policy"?

I have updated the description. Please let me know if you would like me to expand more?

**Comment by trapexit** [comment]
Why that particular algo with the random value between 0 and the sum of all drives? Is that materially better than just using `rand` which seems very similar to me. Filter out those ineligible and pick one at random.

**Comment by andreasDalsgaard** [comment]
It is essentially an efficient way which avoid constructing data structures otherwise commonly constructed for alias method algorithms. Yes, with rand we would have 15 files on source A and 15 files on source B.

**Comment by trapexit** [comment]
You might want to keep the fs info as 1) the call is a bit costly when statvfs caching is off. 2) more importantly... there is a race condition which could result in nothing being selected. Files could be deleted causing the sum to be lower than the previously calculated total. If the random value (why do you call it "byte"?) is higher than the new, lower total it'll fail to select anything.

I'm fine with the general idea but I think the name needs work. Something like "pfrd"? Percentage Free Random Distribution?

**Comment by andreasDalsgaard** [comment]
Ah, I thought "rwlock::ReadGuard guard(&branches_.lock);" was a lock preventing that. I have made a cache with a fixed size. However, I wanted to hear your view on stack allocation? E.g. alloca may not be available on some architectures

**Comment by trapexit** [comment]
Prevent what? The lock is locking branches structure. It won't change over the life of it being held but the size of the structure is not fixed. It's unlikely someone would have such a large branch list but that's besides the point.

Stack allocation is probably fine but not sure it's worth it. It's not like there is a no-allocation thing going on here. I'd just use a std::vector of a custom struct to hold the info needed.

File: src/policy_pfrd.cpp [hunk 1]
```diff
@@ -51,10 +56,9 @@ namespace pfrd
     fs::info_t info;
     const Branch *branch;
     const string *pfrdbasepath;
-    uint64_t cache_spaceavail_branches[256];
-    bool useable_branches[256];
-    const string *cache_basepath_branches[256];
     size_t branch_count = branches_.size();
+    bool useable_branches[branch_count];
```
Comment (trapexit) [inline src/policy_pfrd.cpp:60]: VLA, as I recall, isn't technically supported in C++ but shouldn't be a big deal so long as it works on gcc and clang. The bool array though... why not use the string ptr for that? Set to NULL at the top of the loop and then in the second loop if == NULL skip it.

Comment (trapexit) [reply src/policy_pfrd.cpp:60]: Also, can you please follow the existing styling? I realize it's non-standard and not currently consistent but I am in the process of fixing that.

Comment (andreasDalsgaard) [reply src/policy_pfrd.cpp]: Yeah, good point!

**Comment by trapexit** [comment]
I've done some misc cleanup that appears to conflict. I can close this and submit it myself or if you want to rework it I can then accept it and do whatever refactor needed.

**Comment by andreasDalsgaard** [comment]
I have merged with master and resolved the conflict. However, feel free to close it and do cleanup/rework as you see fit.

**Comment by trapexit** [comment]
https://github.com/trapexit/mergerfs/tree/policy-pfrd

How's this look? Created 3 new policies. pfrd, eppfrd, and msppfrd.

**Comment by trapexit** [comment]
@andreasDalsgaard Will you get a chance to look at my patch?

**Comment by andreasDalsgaard** [comment]
@trapexit I wrote a comment here: https://github.com/trapexit/mergerfs/commit/8a5579385d8543ec58bcff27737a917153138da3

**Pull request closed without merging**